test_runner/regress/test_pageserver_generations.py

```diff
@@ -0,0 +1,352 @@
+"""
+
+Tests in this module exercise the pageserver's behavior around generation numbers,
+as defined in docs/rfcs/025-generation-numbers.md.  Briefly, the behaviors we require
+of the pageserver are:
+- Do not start a tenant without a generation number if control_plane_api is set
+- Remote objects must be suffixed with generation
+- Deletions may only be executed after validating generation
+- Updates to remote_consistent_lsn may only be made visible after validating generation
+"""
+
+
+import re
+import time
+from typing import Optional
+
+import pytest
+from fixtures.log_helper import log
+from fixtures.neon_fixtures import (
+    NeonEnv,
+    NeonEnvBuilder,
+    PgBin,
+    last_flush_lsn_upload,
+    wait_for_last_flush_lsn,
+)
+from fixtures.pageserver.utils import list_prefix
+from fixtures.remote_storage import (
+    RemoteStorageKind,
+)
+from fixtures.types import TenantId, TimelineId
+from fixtures.utils import print_gc_result, wait_until
+
+# A tenant configuration that is convenient for generating uploads and deletions
+# without a large amount of postgres traffic.
+TENANT_CONF = {
+    # small checkpointing and compaction targets to ensure we generate many upload operations
+    "checkpoint_distance": f"{128 * 1024}",
+    "compaction_threshold": "1",
+    "compaction_target_size": f"{128 * 1024}",
+    # no PITR horizon, we specify the horizon when we request on-demand GC
+    "pitr_interval": "0s",
+    # disable background compaction and GC. We invoke it manually when we want it to happen.
+    "gc_period": "0s",
+    "compaction_period": "0s",
+    # create image layers eagerly, so that GC can remove some layers
+    "image_creation_threshold": "1",
+}
+
+
+def generate_uploads_and_deletions(
+    env: NeonEnv,
+    *,
```

Comment:
I am not familiar with this python-fu, what does the star do here?

Reply:
Keyword-only arguments: https://python-3-for-scientists.readthedocs.io/en/latest/python3_advanced.html#keyword-only-arguments

It makes python code easier to refactor (and in some cases more robust) when adding optional arguments, by forcing callers to use keyword-style invocation rather than positional.  Usually, when we add arguments to a function we have to add them to the end, or carefully edit all existing call sites.  When using keyword-only arguments, we can add/reorder anything after the `*` safely.

They're not appropriate for smaller helper functions where terseness is important, but I find them quite useful on functions that have some more rarely used optional arguments.

Reply:
> keyword-only arguments
oki, that I was expecting, but wasn't sure. I quite dislike that language feature though :)